patch 2 Pavel method

Jabkin1 opened the pull request:
Pavel's method for image processing in image processing.py and its import in label_weeds.py

Jabkin1 committed:
patch 2 Pavel method

Pavel's method for image processing in image processing.py and its import in label_weeds.py

tajgr commented:
This PR is in conflict with previous #5 . I recommend merge them.

Jabkin1 committed:
Json format output update

Adds VGG json format output + palevs method implementation

Jabkin1 committed:
update 3 auto annotating button 

Including the auto annotation button for "o" which annotates all objects in the image to the current class.

tajgr commented on label_weeds.py:
```diff
@@ -58,13 +58,15 @@ def make_hist_im(self, gray):
         nir = transform_nir(im_nir)
-        if method == "ndvi":
+        if method == "pavel":
+            gray == pavel_method(im_color, nir)
```
Patrně překlep.

tajgr commented on label_weeds.py:
```diff
@@ -80,16 +82,66 @@ def make_gray_im(self, color_path, method="ndvi"):
+            if item.endswith("rgb.tif"):
                 color_path = os.path.join(self.path, item)
-                nir_path = os.path.join(self.path, item[:-5]+"N.tif")
+                nir_path = os.path.join(self.path, item[:-5]+"nir.tif")
```
Místo -5 je potřeba dát -7.

Jabkin1 committed:
Revision of errors

Updated code with the suggested updates implemented

Jabkin1 committed:
Update 5 simplified method + anotations name input

The method is simplified and the import of images was adjusted for the current naming convention. Added input for annotation names and active switching between them in the windows.

Jabkin1 committed:
label_weeds Modify morphology for bbox

Jabkin1 committed:
Pascal VOC annotations and output changes

major update including a new annotation output format

Jabkin1 committed:
Output annotations showcase

A few examples of annotations and the used version of Labelimg

tajgr commented:
Obecně je dobré do PR napsat, jak se s kódem pracuje.

tajgr commented on image_processing.py:
```diff
@@ -43,6 +43,21 @@ def get_ndvi_im(im_color, nir):
+    processed_im = (((2 * nir + g - r - b) / (2 * nir + g + r + b + 510) - 0.05) * 1.8 + 1) * 128.0
+    # processed_im = cv2.convertScaleAbs(processed_im)
+    # processed_im = (processed_im - processed_im.min())/(processed_im.max() - processed_im.min()) * 255
+    # processed_im = (processed_im - 90) / (170 - 90) * 255
```
Je potřeba udržovat zakomentované staré verze?

tajgr commented on image_processing.py:
```diff
@@ -168,3 +183,4 @@ def test_combination(path):
         test_excess_green(args.path)
     if args.com:
         test_combination(args.path)
+        
```
Poslední řádka.

tajgr commented on label_weeds.py:
```diff
@@ -1,35 +1,24 @@
-        self.current_label = "0"
+        assert os.path.isdir(out), f"Output directory '{out}' does not exist."
+        self.out_dir = out
+        self.labels = [str(label) for label in label_names]  # Convert labels to strings
```
Co když label_names je prázdný list?

tajgr commented on label_weeds.py:
```diff
@@ -74,38 +65,83 @@ def make_gray_im(self, color_path, method="ndvi"):
+
+            tree = ET.ElementTree(root_elem)
+            xml_path = os.path.join(self.out_dir, f"{os.path.splitext(os.path.basename(image_name))[0]}.xml")
+            tree.write(xml_path, encoding="utf-8", xml_declaration=True)
```
Ten xml jde určitě zapsat nějak lidštěji, tj. ne vše do jedné řádky.. né že bych věděl jak, ale podíval bych se na to.

tajgr replied:
minidom?
https://docs.python.org/3.10/library/xml.dom.minidom.html

tajgr commented on label_weeds.py:
```diff
@@ -261,8 +326,9 @@ def run(self):
     parser = argparse.ArgumentParser(description='__doc__')
     parser.add_argument('path', help='Path to image directory.')
-    parser.add_argument('--out', help='Specify an annotation file name (.json)', default="test.json")
+    parser.add_argument('--out', help='Specify an output directory for annotations.', required=True)
```
V helpu by asi mělo být info, že jde o povinný argument, případně default na `path`?

tajgr commented on label_weeds.py:
```diff
@@ -261,8 +326,9 @@ def run(self):
     import argparse
     parser = argparse.ArgumentParser(description='__doc__')
```
Tisknutí `"__doc__"` nemá smysl. Tady má být `descripption`, které by bylo v `__doc__`, jenže to jsi zrušil.

tajgr commented on label_weeds.py:
```diff
@@ -261,8 +326,9 @@ def run(self):
     parser.add_argument('path', help='Path to image directory.')
-    parser.add_argument('--out', help='Specify an annotation file name (.json)', default="test.json")
+    parser.add_argument('--out', help='Specify an output directory for annotations.', required=True)
+    parser.add_argument('--labels', nargs='+', help='List of label names.')
```
Nastavil bych default hodnoty pro labels

tajgr replied:
LabelImg myslím požaduje labels v samostatném texťáku. Nechceš si ten soubor vytvořit dopředu a zde automaticky načítat? Jakákoli ruční práce je zdrojem chyb.

Jabkin1 replied:
Stačí momentálně jen jeden druh na masovou anotaci.

tajgr commented:
U labelImg jsem narazil na toto issue: https://github.com/HumanSignal/labelImg/issues/938

tajgr commented on label_weeds.py:
```diff
@@ -1,35 +1,24 @@
-                json.dump(self.data, outfile, indent=4)
-        else:
-            self.data = {}
+        self.data = {}
```
Tj. již se nejde vrátit k původní práci. Je to tak správně?

tajgr commented on label_weeds.py:
```diff
@@ -74,38 +65,83 @@ def make_gray_im(self, color_path, method="ndvi"):
+            tree.write(xml_path, encoding="utf-8", xml_declaration=True)
+
+    @staticmethod
+    def make_object_element(bbox, label):
```
Statické metody mám rád pohromadě hned za __init__

tajgr commented on label_weeds.py:
```diff
@@ -74,38 +65,83 @@ def make_gray_im(self, color_path, method="ndvi"):
         file_list = os.listdir(self.path)
         for item in sorted(file_list):
-            if item.endswith("C.tif"):
+            if item.lower().endswith("_rgb.tiff"):
```
O tomhle jsme se bavily.. je to divný.

Jabkin1 replied:
Chová se to divně, ale nedokážu potřebu .lower vysvětlit

tajgr commented on label_weeds.py:
```diff
@@ -223,35 +284,39 @@ def run(self):
-                self.current_label = chr(k)
-
-            elif k in [0x53, ord("d")]:  # take next image
+            elif k == ord("o"):  # Annotate all objects with the current label
```
pokud spleteš label, můžeš resetovat celý obrázek najednou? Myslím že ne.

tajgr commented on showcase/im_037676_rgb.xml:
```diff
@@ -0,0 +1,2 @@
+<?xml version='1.0' encoding='utf-8'?>
+<annotation><filename>im_037676_rgb.tiff</filename><object><name>parph</name><bndbox><xmin>369</xmin><ymin>1391</ymin><xmax>391</xmax><ymax>1405</ymax></bndbox></object><object><name>parph</name><bndbox><xmin>438</xmin><ymin>1260</ymin><xmax>479</xmax><ymax>1292</ymax></bndbox></object><object><name>parph</name><bndbox><xmin>1762</xmin><ymin>1198</ymin><xmax>1779</xmax><ymax>1219</ymax></bndbox>< ...
```
Je to slité do jednoho řádku a ještě nekorektně ukončené.

tajgr replied:
Chceme ty xml v gitu? Chápu, chtěl jsem tu vzorové data, ale i tak bych tu nechal jen obrázky. Možná, kdyby to byla součást unittestu.

tajgr replied:
Ty xml dej pryč, nebo alespoň přegeneruj.

tajgr commented on showcase/info:
```diff
@@ -0,0 +1,2 @@
+Output images in Pascal VOC fomat, compatible with labelImg 1.8.6 (2021-10-10).
```
Tohle patří do `__doc__` v label_weeds.py.

tajgr commented:
Přidej prosím ještě `requirements.txt`. Ať máme sjednocené verze knihoven. Dík.

Jabkin1 committed:
Final version resolving xml format, updating the issues

Jabkin1 committed:
Finally updated for better xml format

tajgr commented on image_processing.py:
```diff
@@ -168,3 +178,5 @@ def test_combination(path):
     if args.com:
         test_combination(args.path)
+
+        
```
Pořád tu vidím chybějící řádku.

tajgr replied:
Aha, ona tam vlastně je, ale není prázdná. Asi jim to vadí...

tajgr reviewed (approved):
Prosím zkontroluj ještě ty řádky na konci souborů a vyřeš ty xml soubory a můžeš mergeovat.

Jabkin1 committed:
Resolve conflicts
